Discovery

theJenix opened the pull request:
New discovery process, complete with multi select dialogs and support for a remote fan to initiate "find new fans".  There is still a bug on the "fan" side when the app has been running a long time, and there's still a question about how we integrate it into the main app (I'm thinking another service maybe), but this is tested working and can be pulled in while we discuss that other stuff.

theJenix committed:
Reworked discovery to scan for devices, accumulate discovered devices, and
present a dialog to the user to select which device to connect to

theJenix committed:
added TODO comment about popping message up to user when connect failed

theJenix committed:
Merge branch 'master' into discovery

Conflicts:
	fanClub/res/values/strings.xml

theJenix committed:
Merge branch 'master' into discovery

theJenix committed:
Added functionality for fans to initiate discovery, and then receive a list of
found fans (fan side NYI).

Started refactoring connection management into a service, to pull it out of any
specific activity.

Moved message handling into a Dispatch class, which will
allow the service or any activities to register handlers for individual
messages.

theJenix committed:
Merge branch 'master' into discovery

theJenix committed:
Merge branch 'master' into discovery

Conflicts:
	fanClub/src/com/lastcrusade/fanclub/util/BluetoothUtils.java

theJenix committed:
Added multi-select support to Found devices dialog, to allow the user to select
multiple fans to add at a time.

Added support for messages to send found fans back to requesting fan, and allow
the fan to send the specific fans to connect back to the host.

theJenix committed:
Added toast message for when fans are found
Cleaned up commented code in PlaybarFragment

theJenix committed:
Added comments

theJenix committed:
Added comments

bryansills commented on fanClub/src/com/lastcrusade/fanclub/HostActivity.java:
```diff
@@ -15,34 +14,49 @@
+/**
+ * 
+ * 
+ * @author Jesse Rosalia
```
now you are just taunting us

theJenix replied:
:laughing:

reidbaker replied:
You mean like this kind of taunting? 
http://www.youtube.com/watch?feature=player_detailpage&v=9V7zbWNznbs#t=156s

ejohnson44 commented on fanClub-test/src/com/lastcrusade/fanclub/net/message/ConnectFansMessageTest.java:
```diff
@@ -0,0 +1,32 @@
+        ConnectFansMessage oldMessage = new ConnectFansMessage(addresses);
+        ConnectFansMessage newMessage = super.testSerializeMessage(oldMessage);
+        //verify that the FoundFan objects are equal and in the same order.
+        for (int ii = 0; ii < addresses.size(); ii++) {
```
Why ii instead of the normal i or index?

theJenix replied:
A habit impressed upon me by my first boss.  ii is much easier to search for within a file than i, as very few words are spelled with 'ii'.

theJenix committed:
Deleted unused list item

bryansills commented on fanClub/src/com/lastcrusade/fanclub/components/MultiSelectListDialog.java:
```diff
@@ -0,0 +1,125 @@
+        };
+    }
+
+    public MultiSelectListDialog<T> setFormatter(
```
Why are you returning the object with a setter?

theJenix replied:
I used the builder pattern with this dialog..the reason you return this from all of those methods is so you can chain methods together.

reidbaker replied:
Is there a place in code that uses the chaining you mention?

theJenix replied:
HostActivity#onDiscoveredDevices

bryansills commented on fanClub/src/com/lastcrusade/fanclub/net/BluetoothDiscoveryHandler.java:
```diff
@@ -0,0 +1,88 @@
+     * @param device
+     */
+    public void onDiscoveryFound(BluetoothDevice device) {
+        Log.w(TAG,
```
Couldn't we just use BluetoothDeviceDialogFormatter.format()?

theJenix replied:
I'm going to say probably not...currently they're the same format, but this is debug output and that is for user presentation.  Those output formats may diverge in the future, and we'd have to change this back.

reidbaker commented on fanClub-test/src/com/lastcrusade/fanclub/net/message/ConnectFansMessageTest.java:
```diff
@@ -0,0 +1,32 @@
+        addresses.add("00:11:22:33:44:55:66");
+        addresses.add("00:11:22:33:44:55:67");
+        addresses.add("00:11:22:33:44:55:68");
+        addresses.add("00:11:22:33:44:55:69");
```
hehe 69

reidbaker commented on fanClub/lint.xml:
```diff
@@ -0,0 +1,3 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Just out of curiosity what is this file for?

theJenix replied:
I'm not entirely sure...Eclipse generated it.  Lint is a static code checker, so maybe this has something to do with packaging the APK or something.  I think we can safely remove/ignore it.

theJenix replied:
It has been safely removed/ignored.

reidbaker commented on fanClub/src/com/lastcrusade/fanclub/FanActivity.java:
```diff
@@ -88,6 +97,30 @@ public void onClick(View v) {
                 onHelloButtonClicked();
```
Does this method exist anymore?

theJenix replied:
It's on line 134.

reidbaker commented on fanClub/src/com/lastcrusade/fanclub/FanActivity.java:
```diff
@@ -88,6 +97,30 @@ public void onClick(View v) {
+            public void handleMessage(int messageNo,
+                    StringMessage message, String fromAddr) {
+                StringMessage sm = (StringMessage)message;
+                Toaster.iToast(FanActivity.this, sm.getString());
```
Is this toast for testing or will it be showing the user important info?

theJenix replied:
This is for handling the string message, which shows "Hello Cleveland" or "You Rock"....I think the Fan gets Hello Cleveland.  It's ultimately a test mechanism, but one I'd like to leave in for now.

reidbaker commented on fanClub/src/com/lastcrusade/fanclub/HostActivity.java:
```diff
@@ -88,48 +103,106 @@ public void onClick(View v) {
+
+            @Override
+            public void handleMessage(int messageNo,
+                    FindNewFansMessage message, String fromAddr) {
```
`fromAddr` is a String but what would be an example of what is in fromAddr?
Is it like a mac or bluetooth name?

theJenix replied:
I added a comment to IMessageHandler to explain the parameters.  It's the bluetooth MAC address.

reidbaker commented on fanClub/src/com/lastcrusade/fanclub/HostActivity.java:
```diff
@@ -144,12 +217,80 @@ protected void onHelloButtonClicked() {
+    /**
+     * Send a message to all connected clients
+     * 
+     * @param sm
```
If we are going to have docstrings lets say what sm is haha

theJenix replied:
Fixed.

reidbaker commented on fanClub/src/com/lastcrusade/fanclub/HostActivity.java:
```diff
@@ -158,111 +299,46 @@ private void broadcastMessage(StringMessage sm) {
+
+        BluetoothAdapter adapter = BluetoothAdapter.getDefaultAdapter();
+        if (adapter == null) {
+            Toaster.eToast(this, "Bluetooth adapter is null");
```
We might want to make this a string in strings.xml.
If it gets shown to the user it should be in an xml file. In fact we may want to have an error.xml with error messages.

theJenix replied:
I removed this check altogether...this method is called to respond to a network message, and onCreate is written to abort if there is no bluetooth, which means that at this point we can assume adapter is nonnull.

reidbaker commented on fanClub/src/com/lastcrusade/fanclub/HostActivity.java:
```diff
@@ -158,111 +299,46 @@ private void broadcastMessage(StringMessage sm) {
-                this.message.append(sm.getString());
-            }
+        if (found == null) {
+            Toaster.eToast(this, "Unknown remote device.");
```
We might want to make this a string in strings.xml.
If it gets shown to the user it should be in an xml file. In fact we may want to have an error.xml with error messages.

theJenix replied:
Changed to a Log.wtf...this was for testing.

reidbaker commented on fanClub/src/com/lastcrusade/fanclub/net/MessageThreadMessageDispatch.java:
```diff
@@ -0,0 +1,66 @@
+import com.lastcrusade.fanclub.net.message.IMessage;
+
+/**
+ * An Android Handler used as a fanClub network message dispatch.  This will receive Android Message objects that represent
```
Long line

theJenix replied:
Fixed

reidbaker commented on fanClub/src/com/lastcrusade/fanclub/net/MessageThreadMessageDispatch.java:
```diff
@@ -0,0 +1,66 @@
+
+    private static final String TAG = "MessageThreadMessageDispatch";
+
+    private Map<Class<? extends IMessage>, IMessageHandler<? extends IMessage>> dispatchMap =
```
Well this line isn' scary. It just needs some more punctuation.

theJenix replied:
I'll add a few more ? and a couple .
And a comment :laughing:

reidbaker commented on fanClub/src/com/lastcrusade/fanclub/net/message/FindNewFansMessage.java:
```diff
@@ -8,9 +8,11 @@
 
     @Override
     public void deserialize(InputStream input) throws IOException {
```
So the fact that we get a message is all we need? If that is true should we do the same thing for play, pause, and skip?

theJenix replied:
Correct.  Greenie and I talked about this yesterday, after he had his messages written, and I suggested they could be left alone but future messages can be done this way.

dgreenhalgh replied:
Yeah, I'll look into this. :horse:

reidbaker commented:
Somewhere there should be a TODO to save the devices we have connected to in the past

theJenix committed:
Removed lint.xml from tree, and added it to .gitignore

theJenix committed:
Added comments to IMessageHandler#handleMessage to explain the parameters

theJenix committed:
Fixed comments in HostActivity

theJenix committed:
Replaced toast message w/ log message in HostActivity
removed superfluous null check (and literal string)
Fixed long line comments in MessageThreadMessageDispatch

theJenix committed:
Merge branch 'master' of git://github.com/TheLastCrusade/fanClub into discovery

Conflicts:
	fanClub/res/values/strings.xml
	fanClub/src/com/lastcrusade/fanclub/util/BluetoothUtils.java

Fixed small formatting issues after merge in BluetoothUtils.java

theJenix commented:
I believe I addressed everything in the review.

dgreenhalgh commented:
I agree. Does anyone have a problem with discovery being pulled in?

reidbaker commented:
I Dont let's pull it in
On Feb 28, 2013 10:19 AM, "David Greenhalgh" notifications@github.com
wrote:

> I agree. Does anyone have a problem with discovery being pulled in?
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/TheLastCrusade/fanClub/pull/29#issuecomment-14238102
> .
